File: java/RandomForest/DecisionTree.java
package RandomForest;

import java.io.*;
import java.text.DecimalFormat;
import java.util.*;

import static RandomForest.util.*;


class TreeNode {
	int feature;
	double value;
	String maxClass;
	double[] probas;
	TreeNode left = null;
	TreeNode right = null;
	
	public TreeNode(int feature, double value) {
		this.feature = feature;
		this.value = value;
	}
	
	public TreeNode(int feature, double value, String maxClass, double[] probas, TreeNode left, TreeNode right) {
		this.feature = feature;
		this.value = value;
		this.maxClass = maxClass;
		this.probas = probas;
		this.left = left;
		this.right = right;
	}
}


class BTree {
	
	TreeNode root;
	
	public BTree(TreeNode root) {
		this.root = root;
	}
	
	public BTree() {}

	public void preOrder(TreeNode node) {
		
		if (node.left == null) {
			System.out.print(node.feature);
			System.out.print("   ");
			System.out.print(node.value);
			System.out.print("   ");
			System.out.println(node.maxClass);
		}
		
		else {
			System.out.print(node.feature);
			System.out.print("   ");
			System.out.print(node.value);
			System.out.print("   ");
			System.out.println(node.maxClass);
			preOrder(node.left);
			preOrder(node.right);
		}
		
	}
	
	public int getNumLeafs(TreeNode node) {
		int numLeafs = 0;
		
		if (node.left == null) {
			numLeafs += 1;
		}
		
		else {
			numLeafs += getNumLeafs(node.left);
			numLeafs += getNumLeafs(node.right);
		}
		
		return numLeafs;
	}
	
	public String getClass(double[] x, TreeNode node) {
		
		if (node.left == null) {
			return node.maxClass;
		}
		
		else {
			if (x[node.feature] > node.value)
				return getClass(x, node.right);
			else
				return getClass(x, node.left);
		}
		
	}
	
	public double[] getProba(double[] x, TreeNode node) {
		
		if (node.left == null) {
			return node.probas;
		}
		
		else {
			if (x[node.feature] > node.value)
				return getProba(x, node.right);
			else
				return getProba(x, node.left);
		}
		
	}
}


public class DecisionTree {
	
	double tol = 0.0001;
	int leastSample = 1;
	int maxDepth = 999;
	String merge = "pep";
	int maxFeatures = 10;
	List<String> targetList = new ArrayList<String>();
	BTree tree = new BTree();
	
	public DecisionTree(double tol, int leastSample, int maxDepth, String merge, int maxFeatures) {
		this.tol = tol;
		this.leastSample = leastSample;
		this.maxDepth = maxDepth;
		this.merge = merge;
		this.maxFeatures = maxFeatures;
	}
	
	public DecisionTree() {
	}
	
	public List binSplitDataSet(List<double[]> data, List<String> y, int feature, double value) {
		List ret = new ArrayList();
		List<double[]> mat0 = new ArrayList<double[]>();
		List<double[]> mat1 = new ArrayList<double[]>();
		List<String> y0 = new ArrayList<String>();
		List<String> y1 = new ArrayList<String>();
		
		for (int i = 0; i < data.size(); i++) {
			double[] row = data.get(i);
			if (row[feature] <= value) {
				mat0.add(row);
				y0.add(y.get(i));
			}
			else {
				mat1.add(row);
				y1.add(y.get(i));
			}
		}
		
		ret.add(mat0); ret.add(y0); ret.add(mat1); ret.add(y1);
		return ret;
	}
	
	public List chooseBestSplit(List<double[]> data, List<String> y) {
		List ret = new ArrayList();
		Set<String> ySet = new HashSet<String>(y);
		
		if (ySet.size() == 1) {
			ret.add(null); ret.add(null); ret.add(y.get(0)); ret.add(probaArray(y));
			return ret;
		}
		
		int n = data.get(0).length;
		double S = gini(y);
		double bestS = 1.0;
		int bestIndex = 0;
		double bestValue = 0.0;
		List<Integer> featureList = new ArrayList<Integer>();
		int visitedFeatures = 0;
		
		for (int i = 0; i < n; i++)
			featureList.add(i);
		
		while (visitedFeatures < maxFeatures && featureList.size() > 0) {
			Random r = new Random();
			int randNum = r.nextInt(featureList.size());
			int featIndex = featureList.get(randNum);
			featureList.remove(randNum);
			List<Double> column = new ArrayList<Double>();
			
			for (double[] i : data) 
				column.add(i[featIndex]);
			
			Set<Double> featSet = new HashSet<Double>(column);
			if (featSet.size() == 1)
				continue;
			
			visitedFeatures += 1;
			/*double lastS = (double) 1.0;*/
			
			for (double splitVal : featSet) {
				List splitted = binSplitDataSet(data, y, featIndex, splitVal);
				List<double[]> mat0 = (ArrayList<double[]>) splitted.get(0);
				List<String> y0 = (ArrayList<String>) splitted.get(1);
				List<double[]> mat1 = (ArrayList<double[]>) splitted.get(2);
				List<String> y1 = (ArrayList<String>) splitted.get(3);
				
				if (mat0.size() < leastSample || mat1.size() < leastSample)
					continue;
				
				double r0 = ((double) y0.size()) / y.size();
				double r1 = ((double) y1.size()) / y.size();
				double newS = r0 * gini(y0) + r1 * gini(y1);
				
				/*if (newS > lastS)
					break;*/
				
				if (newS < bestS) {
					bestIndex = featIndex;
					bestValue = splitVal;
					bestS = newS;
				}
				
				/*lastS = newS;*/
			}
		}
		
		if (S - bestS < tol) {
			ret.add(null); ret.add(null); ret.add(classify(y)); ret.add(probaArray(y));
			return ret;
		}
		
		List splitted = binSplitDataSet(data, y, bestIndex, bestValue);
		List<double[]> mat0 = (ArrayList<double[]>) splitted.get(0);
		List<String> y0 = (ArrayList<String>) splitted.get(1);
		List<double[]> mat1 = (ArrayList<double[]>) splitted.get(2);
		List<String> y1 = (ArrayList<String>) splitted.get(3);
		
		if (mat0.size() < leastSample || mat1.size() < leastSample) {
			ret.add(null); ret.add(null); ret.add(classify(y)); ret.add(probaArray(y));
			return ret;
		}
		
		ret.add(bestIndex); ret.add(bestValue); ret.add(classify(y)); ret.add(probaArray(y));
		return ret;
	}
	
	public TreeNode createTree(List<double[]> data, List<String> y, int depth) {
		TreeNode retTree = new TreeNode(-1, -1);
		
		if (depth > maxDepth) {
			retTree.maxClass = classify(y);
			retTree.probas = probaArray(y);
			return retTree;
		}
		
		List bestSplit = chooseBestSplit(data, y);
		Object feat = bestSplit.get(0); Object val = bestSplit.get(1); String mc = (String) bestSplit.get(2); double[] probas = (double[]) bestSplit.get(3);
		if (feat == null) {
			retTree.maxClass = mc;
			retTree.probas = probas;
			return retTree;
		}
		
		retTree.feature = (int) feat;
		retTree.value = (double) val;
		retTree.maxClass = mc;
		retTree.probas = probas;
		List splitted = binSplitDataSet(data, y,  (int) feat, (double) val);
		List<double[]> left = (ArrayList<double[]>) splitted.get(0);
		List<String> yleft = (ArrayList<String>) splitted.get(1);
		List<double[]> right = (ArrayList<double[]>) splitted.get(2);
		List<String> yright = (ArrayList<String>) splitted.get(3);
		retTree.left = createTree(left, yleft, depth + 1);
		retTree.right = createTree(right, yright, depth + 1);
		
		return retTree;
	}
	
	public boolean isLeaf(TreeNode node) {
		return (node.left == null);
	}
	
	private TreeNode prune_pep(TreeNode node, List<double[]> data, List<String> y) {
		if (isLeaf(node))
			return node;
		
		int numLeafs = tree.getNumLeafs(node);
		long length = y.size();
		double errorNoMerge = (1 - score(data, y)) * length + 0.5 * numLeafs;
		double varNoMerge = Math.sqrt(errorNoMerge * (1 - errorNoMerge / length));
		List<String> yPred = new ArrayList<String>();
		
		for (int i = 0; i < length; i++) {
			yPred.add(node.maxClass);
		}
		
		double errorMerge = length - correctNum(y, yPred) + 0.5;
		
		if (errorMerge < errorNoMerge + varNoMerge) {
			node.feature = -1; node.value = -1; node.left = null; node.right = null;
			return node;
		}
		
		else {
			List splitted = binSplitDataSet(data, y, node.feature, node.value);
			List<double[]> left = (ArrayList<double[]>) splitted.get(0);
			List<String> yleft = (ArrayList<String>) splitted.get(1);
			List<double[]> right = (ArrayList<double[]>) splitted.get(2);
			List<String> yright = (ArrayList<String>) splitted.get(3);
			
			if (!isLeaf(node.left))
				node.left = prune_pep(node.left, left, yleft);
			
			if (!isLeaf(node.right))
				node.right = prune_pep(node.right, right, yright);
			
			return node;
		}
	}
	
	public void train(List<double[]> data, List<String> y) {
		Set<String> targetSet = new HashSet<String>(y);
		targetList = new ArrayList<String>(targetSet);
		Collections.sort(targetList);
		
		long length = y.size();
		long cutPoint = length * 9 / 10;
		List<String> ytrain = new ArrayList<String>();
		List<double[]> train = new ArrayList<double[]>();
		List<String> ytest = new ArrayList<String>();
		List<double[]> test = new ArrayList<double[]>();
		
		for (int i = 0; i < data.size(); i ++) {
			if (i < cutPoint) {
				train.add(data.get(i));
				ytrain.add(y.get(i));
			}
			
			else {
				test.add(data.get(i));
				ytest.add(y.get(i));
			}
		}
		
		if (merge == "None") {
			TreeNode root = createTree(data, y, 0);
			tree = new BTree(root);
		}
		
		else if (merge == "pep" ) {
			TreeNode root = createTree(data, y, 0);
			tree = new BTree(root);
			root = prune_pep(root, data, y);
			tree = new BTree(root);
		}
		
		else
			System.out.println("Prune type does not exist!");
	}
	
	public List predict(List<double[]> data) {
		List<String> ret = new ArrayList<String>();
		
		for (double[] x : data) {
			String pred = tree.getClass(x, tree.root);
			ret.add(pred);
		}
		
		return ret;
	}
	
	public double[] probaArray(List<String> y) {
		DecimalFormat df = new DecimalFormat( "0.00000 ");
		double[] ret = new double[targetList.size()];
		
		for (int i = 0; i < ret.length; i ++) 
			ret[i] = (double) 0.0;
		
		for (String item : y) {
			int index = targetList.indexOf(item);
			ret[index] += 1;
		}
		
		for (int i = 0; i < ret.length; i ++) {
			ret[i] /= y.size();
			ret[i] = Double.parseDouble(df.format(ret[i]));
		}
		
		return ret;
	}
	
	public List<double[]> predictProba(List<double[]> data) {
		List<double[]> ret = new ArrayList<double[]>();
		
		for (double[] x : data) {
			double[] pred = tree.getProba(x, tree.root).clone();
			ret.add(pred);
		}
		
		return ret;
	}
	
	public double score(List<double[]> data, List<String> y) {
		List<String> yPred = new ArrayList<String>();
		yPred = predict(data);
		return correctNum(y, yPred) / ( (double) y.size() );
	}
	
	public static void main(String[] args) throws IOException  {
		String filename = "/home/yejiming/desktop/python/ML_Algorithms/dataset.csv";
		List ret = loadData(filename);
		List<double[]> data = (List<double[]>) ret.get(0);
		List<String> y = (List<String>) ret.get(1);
		List<String> ytrain = new ArrayList<String>();
		List<double[]> train = new ArrayList<double[]>();
		List<String> ytest = new ArrayList<String>();
		List<double[]> test = new ArrayList<double[]>();
		
		for (int i = 0; i < data.size(); i ++) {
			if (i < 50000) {
				train.add(data.get(i));
				ytrain.add(y.get(i));
			}
			
			else {
				test.add(data.get(i));
				ytest.add(y.get(i));
			}
		}
		
		DecisionTree learner = new DecisionTree(0.0001, 1, 999, "pep", 10);
		learner.train(train, ytrain);
		List<double[]> yp = learner.predictProba(test);
		for (int i = 0; i < yp.size(); i++) {
			double[] p1 = yp.get(i);
			for (int j = 0; j < p1.length; j++) {
				p1[j] += 1;
			}
		}
		double[] p = yp.get(0);
		for (double i : p)
			System.out.println(i);
	}
}

File: python/RandomForest/DecisionTree.py
import pandas as pd
from numpy import *
import numpy as np
import random
from math import sqrt
from sklearn import preprocessing
import time
from operator import itemgetter

from utils import *


class TreeNode(object):
    def __init__(self, feat, val, mc, probas, left = 0, right = 0):
        self.feature = feat
        self.value = val
        self.maxClass = mc
        self.left = left
        self.right = right
        self.probas = probas


class BTree(object):
    def __init__(self, root = 0):
        self.root = root
        
    def is_empty(self):
        if self.root is 0:
            return True

        else:
            return False
        
    def preOrder(self, treenode):
        if isinstance(treenode, str) or isinstance(treenode, int):
            return treenode

        print treenode.feature, treenode.value, treenode.maxClass
        self.preOrder(treenode.left)
        self.preOrder(treenode.right)
        
    def inOrder(self, treenode):
        if isinstance(treenode, str) or isinstance(treenode, int):
            return treenode

        self.inOrder(treenode.left)
        print treenode.feature, treenode.value, treenode.maxClass
        self.inOrder(treenode.right)
        
    def postOrder(self, treenode):
        if isinstance(treenode, str) or isinstance(treenode, int):
            return treenode

        self.postOrder(treenode.left)
        self.postOrder(treenode.right)
        print treenode.feature, treenode.value, treenode.maxClass

    def getNumLeafs(self, treenode):
        numLeafs = 0

        if treenode.left == 0 and treenode.right == 0:
            numLeafs += 1

        else:
            numLeafs += self.getNumLeafs(treenode.left)
            numLeafs += self.getNumLeafs(treenode.right)

        return numLeafs

    def getClass(self, x, treenode):
        if treenode.left == 0 and treenode.right == 0:
            return treenode.maxClass
        
        else:
            if x[treenode.feature] > treenode.value:
                return self.getClass(x, treenode.right)
            
            else:
                return self.getClass(x, treenode.left)

    def getProba(self, x, treenode):
        if treenode.left == 0 and treenode.right == 0:
            return treenode.probas
        
        else:
            if x[treenode.feature] > treenode.value:
                return self.getProba(x, treenode.right)
            
            else:
                return self.getProba(x, treenode.left)

        
class ClassificationTree(object):

    def __init__(self, tol=0.0001, leastSample=4,
                 maxDepth=inf, merge=None, maxFeatures=10):
        self.tol = tol
        self.leastSample = leastSample
        self.tree = BTree()
        self.maxDepth = maxDepth
        self.merge = merge
        self.maxFeatures = maxFeatures

    def chooseBestSplit(self, data, y, tol, leastSample):
        tol = self.tol
        leastSample = self.leastSample
        maxFeatures = self.maxFeatures
        
        if len(set(y)) == 1:
            return None, None, y[0], {y[0]: 1.0}
        
        m, n = shape(data)
        S = gini(y)
        bestS = inf
        bestIndex = 0
        bestValue = 0
        featureList = range(n)

        visitedFeatures = 0

        while (visitedFeatures < maxFeatures and len(featureList) >= 1):
            featIndex = random.sample(featureList, 1)[0]
            featureList.remove(featIndex)
            featSet = set(data[:, featIndex])
            lastnewS = inf
            
            if len(featSet) == 1:
                continue

            visitedFeatures += 1

            # Build the dictionary of y for each feature value
            # start = time.time()
            xd, yd = feat_map(data, y, featIndex, featSet)
            # print time.time() - start

            # whole = sorted(zip(data, y), key=itemgetter(featIndex), reverse=False)
            # data = whole[:, :-1]
            # y = whole[: -1]

            # Make the loop to find the best split value
            for split_val in featSet:
                # mat0, y0, mat1, y1 = bin_split(data, y, featIndex, split_val)
                length1 = length2 = 0
                for key in yd:
                    if int(key) <= split_val:
                        length1 += yd[key]
                    else:
                        length2 += yd[key]

                if (length1 < leastSample) or (length2 < leastSample):
                    continue
                
                r0 = float(length1) / len(y)
                r1 = float(length2) / len(y)

                d1 = xd[split_val][:]
                d2 = [0.0 for i in range(9)]
                for key in xd:
                    class_list = xd[key]
                    if int(key) < split_val:
                        for t, i in enumerate(class_list):
                            d1[t] += i
                    elif int(key) > split_val:
                        for t, i in enumerate(class_list):
                            d2[t] += i

                g1 = calc_gini(d1, length1)
                g2 = calc_gini(d2, length2)
                newS = r0 * g1 + r1 * g2

                if newS > lastnewS:
                    break
                
                if newS < bestS:
                    bestIndex = featIndex
                    bestValue = split_val
                    bestS = newS
                    
                lastnewS = newS

        if (S - bestS) < tol:
            return None, None, classify(y), probaDict(y)
        
        mat0, y0, mat1, y1 = bin_split(data, y, bestIndex, bestValue)

        if (shape(mat0)[0] < leastSample) or (shape(mat1)[0] < leastSample):
            return None, None, classify(y), probaDict(y)

        return bestIndex, bestValue, classify(y), probaDict(y)
        
    def createTree(self, data, y, tol, leastSample, depth = 0):
        tol = self.tol
        leastSample = self.leastSample
        retTree = TreeNode(-1, -1, -1, -1)
        
        if depth >= self.maxDepth:
            retTree.feature = None
            retTree.value = None
            retTree.maxClass = self.classify(y)
            retTree.probas = probaDict(y)
            return retTree
        
        feat, val, mc, probas = self.chooseBestSplit(data, y, tol, leastSample)
        if feat == None:
            retTree.feature = None
            retTree.value = None
            retTree.maxClass = mc
            retTree.probas = probas
            return retTree
        
        retTree.feature = feat
        retTree.value = val
        retTree.maxClass = mc
        retTree.probas = probas
        left, yleft, right, yright = bin_split(data, y, feat, val)
        retTree.left = self.createTree(left, yleft, tol, leastSample, depth + 1)
        retTree.right = self.createTree(right, yright, tol, leastSample, depth + 1)

        return retTree

    def train(self, data, y):
        tol = self.tol
        leastSample = self.leastSample
        length = len(y)
        train = data[:(length/10)*9]
        test = data[(length/10)*9:]
        ytrain = y[:(length/10)*9]
        ytest = y[(length/10)*9:]
        
        if self.merge == 'pep':
            root = self.createTree(data, y, tol, leastSample)
            self.tree = BTree(root)
            root = self.prune_pep(root, data, y)
            
        if self.merge == None:
            root = self.createTree(data, y, tol, leastSample)
            
        self.tree = BTree(root)

    def isLeaf(self, treenode):
        return (treenode.feature == None)

    def prune_pep(self, treenode, data, y):
        if self.isLeaf(treenode):
            return treenode
        
        numLeafs = self.tree.getNumLeafs(treenode)
        errorNoMerge = (1 - self.score(data, y)) * len(y) + 0.5 * numLeafs
        varNoMerge = sqrt(errorNoMerge * (1 - errorNoMerge / float(len(y))))
        y_pred = [treenode.maxClass] * len(y)
        errorMerge = len(y) - correctNum(y_pred, y) + 0.5
        
        if errorMerge < errorNoMerge + varNoMerge:
            treenode.feature = None
            treenode.value = None
            treenode.left = 0
            treenode.right = 0
            return treenode
        
        else:
            left, yleft, right, yright = bin_split(data, y,
                                            treenode.feature, treenode.value)

            if not self.isLeaf(treenode.left):
                treenode.left = self.prune_pep(treenode.left, left, yleft)

            if not self.isLeaf(treenode.right):
                treenode.right = self.prune_pep(treenode.right, right, yright)

            return treenode

    def predict(self, data):
        bt = self.tree
        ret = []
        
        for x in data:
            y_pred = bt.getClass(x, bt.root)
            ret.append(y_pred)
            
        ret = array(ret)
        return ret

    def predict_proba(self, data):
        bt = self.tree
        ret = []
        
        for x in data:
            probas = bt.getProba(x, bt.root)
            ret.append(probas)
            
        ret = array(ret)
        return ret

    def score(self, data, y):
        y_pred = self.predict(data)
        return float(correctNum(y, y_pred)) / len(y)
        

def calc_gini(d, total):
    to_square = list()
    for i in d:
        to_square.append(i/float(total))
    to_square = np.array(to_square)
    return 1 - (to_square * to_square.T).sum()

if __name__ == '__main__':
    filename = "dataset.csv"
    df = pd.read_csv(filename, header = 0)
    data = df.values
    y = data[:, -1]
    lbl_enc = preprocessing.LabelEncoder()
    y = lbl_enc.fit_transform(y)
    data = data[:, 0:-1]
    train = data[0:50000]
    ytrain = y[0:50000]
    test = data[50000:60000]
    ytest = y[50000:60000]
    learner = ClassificationTree()
    learner.train(train, ytrain)
    s = learner.score(test, ytest)
    print s
